### ISSUE TITLE
References to cells won't get updated after inserting or deleting cells inside the same file.

### ISSUE THREAD

**Issue by piiskop** [ONLYOFFICE/DocumentServer#942]
**Do you want to request a *feature* or report a *bug*?**
Bug.
**What is the current behavior?**
The references to cells won't get updated after removing or inserting cells.
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
1. Create a spreadsheet!
2. Insert some numbers into first column!
3. Into the second column, into a cell create a "SUM"-formula that covers the entered numbers!
4. Insert a row before the first row of numbers!
5. See that the sum isn't correct anymore as the formula still contains the old references although the numbers have been moved by one row down!
**What is the expected behavior?**
That the references of all the possible formulas in the file will be updated on every insertion or deletion of cells inside that file.
**DocumentServer version:**
5.4.2.46
**Operating System:**
"Ubuntu"
**Browser version:**
"Chrome 84.0.4147.89 (Official Build) (64-bit)"

**Comment by ShockwaveNN** [comment]
1. I still do not fully understand scenario in which you get data loss. Since you can be sure that version of product you used before you upgrade to 5.6.0
2. Some file can be at `/var/lib/onlyoffice/documentserver/App_Data/cache`
3. Dependency problem - `postgresql and (mariadb or mysql)` - yeah I think this is something not right, but better to create new issue with this problem alone

**Comment by piiskop** [comment]
1. I managed to copy most of the lost data into new files as I had these files during the updates open for many days in one of my laptops.
2. There only seem to be some DOCX- and _Editor.bin_-files and nothing for the mentioned three days, only files before and after. _Editor.bin_-files are binary, so I can't read out any data for my spreadsheet files.
3. I'm going to do that.

**Comment by ShockwaveNN** [comment]
> I had these files during the updates open for many days in one of my laptops.

Did you run `documentserver-prepare4shutdown.sh` before upgrading DocumentServer as stated in [upgrade guide](https://helpcenter.onlyoffice.com/server/linux/document/upgrade-document.aspx)

If not - this is a reason why you lost some data during update

**Comment by piiskop** [comment]
No, I didn't. I thought that changes were stored in user files themselves not somewhere in the database or cache as I read from somewhere where it was stated that the database was not being used for file changes but only useful for something like logs.

In your linked upgrade guide, there is:

> Starting from version 4.1, Document Server uses PostgreSQL instead of MySQL database which had been used in previous versions, so in this guide we will show the upgrade from version 4.1 to the latest available version.

This confirms your statement that only _PostgreSQL_ is in use and my doubt that the dependency for _MySQL_ or _mariadb_ isn't needed anymore unless for backwards compatibility probably.

Here, I also read another important rule:

> Due to the operational characteristic, Document Server saves a document only after the document has been closed by all the users who edited it.

Why haven't I seen this rule before? Is it possible that it's only written in the upgrade guide and not in a normal user guide? I've also been looking for a comprehensive user guide and not found any. There is a specific spreadsheets user guide that I discovered by accidentally pressing _F1_ instead of _1_ once a spreadsheet was open though.

Last night, I spent some time on looking for a solution how to use regular expression functions in ONLYOFFICE. It turned out that they aren't available at all! There's no function for extracting nor splitting text! How can that be? Maybe I was a Google Spreadsheets user for too long. Some of my worksheets use regular expressions heavily and I can't use them in ONLYOFFICE. Is this correct or is there an option for regular expressions? I've found we can create macros but this requires rather programming skills from users.

This rule takes time to get used to. Let's imagine a scenario that some people keep a file open for days or weeks like I do! This means that nobody can download the file with the changes done during these days or weeks. Such an anomaly doesn't exist in Google Drive. What if somebody who has no access to that kick-out-script needs to download the file that's in use by like ten other people out of eight aren't reachable? Why couldn't there be a real-time save?

In Google Drive, I can undo and redo independently of how many people are currently editing the same file. In ONLYOFFICE, I can only do it in _strict mode_. And I need to remember save the changes which I never had to do in Google Drive.

**Comment by ShockwaveNN** [comment]
> Is it possible that it's only written in the upgrade guide and not in a normal user guide

What is normal user guide by your definition? This is our official guide on upgrading

I'm sorry, but I think you somehow missing point of Github Issue repo.
This is not a place to discuss some problems. This is place to report issues.

I see an issue with DB dependency.

> Let's imagine a scenario that some people keep a file open for days or weeks like I do!

We make it that way because of technical limitation, we'll hope to fix it, and have some working results with `forceSave` option, but this is not work in all possbile scenarios

> Last night, I spent some time on looking for a solution how to use regular expression functions in ONLYOFFICE

Please do not make this issue a place to report all your troubles. This is Github Issue repo, not some kind of forum.
If you missing something - please report an issue, single issue for EACH feature

**Comment by piiskop** [comment]
A normal user guide for me is a guide that any user of the product can read and that covers all the functionalities of the product. This is a guide only for upgrade. This information should be available in the guide that covers thoughoully all the functions of ONLYOFFICE because it's important to know that if I don't close the file then nobody can download changes made meanwhile. And still: what if somebody never closes a file?

I know that this is a place for reporting issues. Our conversation went viral from a real issue that particularly seems to be solved by now however many more issues have been discussed here. Of course, we could go into another room and leave this room alone. Usually, I report different issues separately but here, the whole story is connected.

Is there an open discussion about the possible scenarios you meant regarding _forceSave_? I found the [process description](https://api.onlyoffice.com/editors/save). If I understand correctly, I can set force saving to be done like every five minutes and in that case, if there are still ten people having the file open their changes all but done during last five minutes are available for download. This API description seems to be pretty useful but still meant for developers not users. And users need this information, too, maybe differently explained.

**Comment by ShockwaveNN** [comment]
> A normal user guide for me is a guide that any user of the product can read and that covers all the functionalities of the product. This is a guide only for upgrade. 

There is plenty of guides on https://helpcenter.onlyoffice.com

> And still: what if somebody never closes a file?

If you run `documentserver-prepare4shutdown.sh` before stopping DocumentServer file will force to close
If documentserver was not stopped - data lost should not happened

> Is there an open discussion about the possible scenarios you meant regarding forceSave? 

Result highly dependant of each integration with DMS you using. It's better be asked in your integration issue tracker
This is not my responsibilites, but i see [several issues](https://github.com/ONLYOFFICE/onlyoffice-nextcloud/issues?q=forcesave) with `forcesave` in onlyoffice-nextcloud repo

**Comment by piiskop** [comment]
On one hand, it's good that all the documentation is gathered in one place. On the other hand, in none of these documentation documents, I can find the information that about how to really do collaboration using spreadsheets and text files besides that API guide that's meant for developers not users. Yes, there is how to do [this and that](https://helpcenter.onlyoffice.com/onlyoffice-editors/index.aspx) but nothing about that saving procedure.

As for _documentserver-prepare4shutdown.sh_, what happens if a file is open and there's a power failure on the server machine? Data loss, right? So if the file has been open for like a week, no changes are actually gathered into the final file that's visible in Nextcloud. After the power failure, if I open that file, I see the state that was a week ago? I want to perform this test by just crashing the ONLYOFFICE server. How can I restart ONLYOFFICE server without restarting the whole machine? As I mentioned before, I installed it using DEB-file, so there's no _snap_ nor _systemctl_ service registered for ONLYOFFICE.

**Comment by ShockwaveNN** [comment]
> in none of these documentation documents, I can find the information that about how to really do collaboration using spreadsheets and text files besides that API guide that's meant for developers not users

Because this depend on integration, each integrator can setup `forcesave` as it want

> How can I restart ONLYOFFICE server without restarting the whole machine

`supervisortctl restart all`

> So if the file has been open for like a week, no changes are actually gathered into the final file that's visible in Nextcloud

Again please not if you experience data loss in that case - please report this problem in integration issue tracker, since integration can have different settings of `forcesave`

**Comment by piiskop** [comment]
I've got so much information and hints from you, thank you!

I did the test and it's like I imagined: in case of a power failure, there's possible data loss. Therefore, I decided to try out _services.CoAuthoring.autoAssembly.interval_ in _/etc/onlyoffice/documentserver/local.js_. And again: the API guide doesn't tell what are the possible values. There's only an example _5m_ for _five minutes_. I'd like to see a fully documented API.

What I did was:

`                        "autoAssembly": {                                                                                                    \

                                "enable": true,
                                "interval": "1m",                                                                                            \

                                "step": "1m"                                                                                                 \

                        },                                                                                                                   \
`
I've no idea what the parameter "step" means. I made a change into a spreadsheet, waited for two minutes, dowloaded the file - change was not present. Maybe I needed to restart the service:
`kalmer@test:/var/www/onlyoffice/documentserver$ sudo supervisorctl restart ds:docservice
ds:docservice: stopped
ds:docservice: started
`
Again, made a change into the spreadsheet, waited for two minutes, downloaded the file from _Nextcloud_. The same result: no change was present. Do you mean that _Nextcloud_'s integration app overrides it somehow?

I made another test. I opened the same file with _Collabora_ and made some changes into the file. They didn't appear in ONLYOFFICE. After that, in fast editing mode, changes in ONLYOFFICE on _ROG_ didn't appear on _ProBook_. Even the red markings weren't available. After I reopened the file, fast mode worked again with red markings.

I just want to be sure that if many people use the same file nobody suffers data loss on power failure of the server machine.

**Comment by ShockwaveNN** [comment]
> Again, made a change into the spreadsheet, waited for two minutes, downloaded the file from Nextcloud. The same result: no change was present. Do you mean that Nextcloud's integration app overrides it somehow?

See https://api.onlyoffice.com/editors/save

There is
> Enable the editorConfig.customization.forcesave mode setting it to true in the editor initialization configuration. In this case each time the user clicks the Save button, the forcesave will be done, and the forcesavetype parameter will have the 1 value when sending the request to the callback handler.

About [that](https://api.onlyoffice.com/editors/config/editor/customization#forcesave) flag

I'm not 100% sure, since as I'm said I'm not very good at nextcloud integration, but

https://github.com/ONLYOFFICE/onlyoffice-nextcloud/search?q=forcesave&unscoped_q=forcesave

this flag is not set via integration, so default value is left which is `false`

**Comment by piiskop** [comment]
I discovered where the bug about updating references could lie. Do it this way:
1. Have a spreadsheet file with at least two worksheets!
2. On the second worksheet, choose a cell and enter a number into that cell!
3. On the first worksheet use _INDIRECT_ function referencing the cell mentioned in the step _2_!
4. On the second worksheet, move the cell somewhere else!
5. On the first worksheet, see that the action performed in the previous step didn't have an influence although the reference in _INDIRECT_ is for the old position. How can that be? That cell's empty now!
6. On the second worksheet, change the content of the cell that has a new position to something that isn't zero!
7. On the first worksheet, see that the cell with _INDIRECT_-formula has the value _0_ because it still references the old position and that cell has nothing inside!

If there are many sheets with _INDIRECT_-references to a same cell, it'd be a pain in the ass as English-speaking people say (pain in the stomach for Germans or head for Estonians) to keep account of all the possible worksheets where the cell is referenced from and update all the references by hand. This is why it can't be a feature but is a bug.

And you don't have to like my tone. But you can like how I find out things that either don't work or are poorly explained.

**Comment by Rita-Bubnova** [comment]
> 1. Have a spreadsheet file with at least two worksheets!
> 2. On the second worksheet, choose a cell and enter a number into that cell!
> 3. On the first worksheet use _INDIRECT_ function referencing the cell mentioned in the step _2_!
> 4. On the second worksheet, move the cell somewhere else!
> 5. On the first worksheet, see that the action performed in the previous step didn't have an influence although the reference in _INDIRECT_ is for the old position. How can that be? That cell's empty now!
> 6. On the second worksheet, change the content of the cell that has a new position to something that isn't zero!
> 7. On the first worksheet, see that the cell with _INDIRECT_-formula has the value _0_ because it still references the old position and that cell has nothing inside!

Sorry, I cannot reproduce your problem with these steps. 
Could you attach file with problem?

**Comment by piiskop** [comment]
I don't know how to attach a file here, so I made a [video](https://youtu.be/alepkpmiyd8). Currently, it's being uploaded.

**Comment by Rita-Bubnova** [comment]
Thank you for the video.

As you yourself have shown, the function `INDIRECT` does not work as you would expect in any editor.  Since XLSX is the main format for us, we calculate functions in the same way to ensure compatibility with MS Excel. This behavior will not change for compatibility with MS Excel.

**Comment by ShockwaveNN** [comment]
This issue was closed due to no response.
